cmd/cstorvolumeclaim/cstorvolumedeployment.go

```diff
@@ -0,0 +1,367 @@
+	}
+	return OwnerReference
+}
+
+// getDeployTemplateAffinity returns affinities
+// for target deployement
+func getDeployTemplateAffinity() *corev1.Affinity {
+	return &corev1.Affinity{
+		PodAffinity: &corev1.PodAffinity{
+			RequiredDuringSchedulingIgnoredDuringExecution: []corev1.PodAffinityTerm{
+				corev1.PodAffinityTerm{
+					LabelSelector: &metav1.LabelSelector{
+						MatchExpressions: []metav1.LabelSelectorRequirement{
+							metav1.LabelSelectorRequirement{
+								Key:      "statefulset.kubernetes.io/pod-name",
+								Operator: metav1.LabelSelectorOpIn,
+								Values:   []string{},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+// getDeployTemplateTolerations returns the array of toleration
+// for target deployement
+func getDeployTemplateTolerations() []corev1.Toleration {
+	return []corev1.Toleration{
+		corev1.Toleration{
+			Effect:            corev1.TaintEffectNoExecute,
+			Key:               "node.alpha.kubernetes.io/notReady",
+			Operator:          corev1.TolerationOpExists,
+			TolerationSeconds: &tolerationSeconds,
+		},
+		corev1.Toleration{
+			Effect:            corev1.TaintEffectNoExecute,
+			Key:               "node.alpha.kubernetes.io/unreachable",
+			Operator:          corev1.TolerationOpExists,
+			TolerationSeconds: &tolerationSeconds,
+		},
+		corev1.Toleration{
+			Effect:            corev1.TaintEffectNoExecute,
+			Key:               "node.kubernetes.io/not-ready",
+			Operator:          corev1.TolerationOpExists,
+			TolerationSeconds: &tolerationSeconds,
+		},
+		corev1.Toleration{
+			Effect:            corev1.TaintEffectNoExecute,
+			Key:               "node.kubernetes.io/unreachable",
+			Operator:          corev1.TolerationOpExists,
+			TolerationSeconds: &tolerationSeconds,
+		},
+	}
+}
+
+func getMonitorMounts() []corev1.VolumeMount {
+	return defaultMounts
+}
+
+func getTargetMgmtMounts() []corev1.VolumeMount {
+	return append(
+		defaultMounts,
+		corev1.VolumeMount{
+			Name:             "tmp",
+			MountPath:        "/tmp",
+			MountPropagation: &mountPropagation,
+		},
+	)
+}
+
+// getDeployTemplateEnvs return the common env required for
+// cstorvolume target deployment
+func getDeployTemplateEnvs(cstorid string) []corev1.EnvVar {
+	return []corev1.EnvVar{
+		corev1.EnvVar{
+			Name:  "OPENEBS_IO_CSTOR_VOLUME_ID",
+			Value: cstorid,
+		},
+		corev1.EnvVar{
+			Name:  "RESYNC_INTERVAL",
+			Value: resyncInterval,
+		},
+		corev1.EnvVar{
+			Name: "NODE_NAME",
+			ValueFrom: &corev1.EnvVarSource{
+				FieldRef: &corev1.ObjectFieldSelector{
+					FieldPath: "spec.nodeName",
+				},
+			},
+		},
+		corev1.EnvVar{
+			Name: "POD_NAME",
+			ValueFrom: &corev1.EnvVarSource{
+				FieldRef: &corev1.ObjectFieldSelector{
+					FieldPath: "metadata.name",
+				},
+			},
+		},
+	}
+}
+
+// getVolumeTargetImage returns Volume target image
+// retrieves the value of the environment variable named
+// by the key.
+func getVolumeTargetImage() string {
```

Comment:
IMO these should be part of builder of the resource that is being built.
We will need to create new builders e.g. `CStorTargetDeploymentBuilder` `CStorTargetServiceBuilder` & so on.
We might probably want to package these structures at:
```go
// pkg/cstor/target/v1alpha1/deployment_builder.go
// pkg/cstor/target/v1alpha1/service_builder.go
```

If we dont use CAST then it should be replaced by specific Builder instance. These may look to be a lot of code, but only by having individual builders & then predicates and so on, we can build the object with finer granularity.


cc @vishnuitta @kmova 

Same comment holds true for all the below functions.

Reply:
Above might need the specific builders to be wrappers over existing builder.
For example, CStorTargetDeploymentBuilder should make use of (wrapper) K8s Deployment Builder.